### DISCUSSION TITLE
How to use all or a specific number of CPU cores?

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#66] [Q&A] [answered]
I have read the help output from `nelua -h` and cannot find anywhere a flag to choose the number of CPU cores I would like to use during compilation and execution time.

Is it something I should done from inside my Nelua code (like a pragma?) or is there a flag / command somewhere which I miss?

**Comment by edubart** [comment, marked as answer]
The Nelua compiler does not use multiple threads to compile at any moment, thus this option does not exist. Writing a multi thread compiler is difficult, time consuming and impose many limitations, moreover the language design decisions like pre processing while compiling does not allow such things happen, because all code must be processed sequentially. The C code is also generated to a single source file for simplicity reasons and other design decisions, thus the C compilation step will also use a single thread because I am not aware of any C compiler that can use multiple threads when compiling a single translation unit.

If you are looking for faster compile time in the C compilation step then use a very fast C compiler like [TCC](https://savannah.nongnu.org/projects/tinycc) with the flag `nelua --cc=tcc`, TCC can compile thousands of C lines in a few milliseconds. If you are looking for faster compilation in the Nelua compiler itself then compile the Nelua's Lua interpreter optimized for your machine with `make optimized-nelua-lua`, this can speedup by ~1.5x factor.

At application execution time, to properly use more cores and threads, you must use a threading library, the [nelua-decl](https://github.com/edubart/nelua-decl) project has bindings for three of them `c89thread` (cross platform), `pthread` (POSIX systems only) and `libuv` (cross platform) has also threading APIs.

**Reply by stefanos82** [reply]
> The Nelua compiler does not use multiple threads to compile at any moment, thus this option does not exist. Writing a multi thread compiler is difficult, time consuming and impose many limitations, moreover the language design decisions like pre processing while compiling does not allow such things happen, because all code must be processed sequentially. The C code is also generated to a single source file for simplicity reasons and other design decisions, thus the C compilation step will also use a single thread because I am not aware of any C compiler that can use multiple threads when compiling a single translation unit.
> 

Yes, you are correct; I never thought of that scenario of sequential processing or preprocessing in general.

> If you are looking for faster compile time in the C compilation step then use a very fast C compiler like [TCC](https://savannah.nongnu.org/projects/tinycc) with the flag `nelua --cc=tcc`, TCC can compile thousands of C lines in a few milliseconds. If you are looking for faster compilation in the Nelua compiler itself then compile the Nelua's Lua interpreter optimized for your machine with `make optimized-nelua-lua`, this can speedup by ~1.5x factor.
>

Interesting the optimization part; I will try it right now. Thank you for letting me know on this one.
 
> At application execution time, to properly use more cores and threads, you must use a threading library, the [nelua-decl](https://github.com/edubart/nelua-decl) project has bindings for three of them `c89thread` (cross platform), `pthread` (POSIX systems only) and `libuv` (cross platform) has also threading APIs.

Here I was thinking more of OpenMP which I have used on a binarytrees demo and executes the code like lightning.

I guess that is what I have had originally in mind...

Is there any example that shows how I could use the `#pragma` in the generated C code before a for loop for heavy computation?

Since you have mentioned TCC, I have tried to use it with the SDL examples and it complains about

```
$ nelua --cc=tcc examples/condots.nelua 
/usr/lib/x86_64-linux-gnu/libpulse.so.0: error: referenced dll 'libpulsecommon-14.2.so' not found
tcc: error: referenced dll 'libpulse.so.0' not found
FPS	13
FPS	13
FPS	13
FPS	13
FPS	13
```
Basically it crawls on its knees.

**Reply by edubart** [reply]
> Is there any example that shows how I could use the #pragma in the generated C code before a for loop for heavy computation?

The condots example uses OpenMP, [check this line](https://github.com/edubart/nelua-lang/blob/master/examples/condots.nelua#L243), it's an example on that. I also do the same [in this ray tracer](https://github.com/edubart/lunray/blob/master/lunray.nelua#L93).

> Basically it crawls on its knees.

TCC doesn't do much optimizations, thus its binary output is usually very inefficient, I recommend using TCC just for quickly prototyping in a live development setup if you think GCC/Clang is being too slow to compile a big project and impacting your productivity.

**Reply by stefanos82** [reply]
> The condots example uses OpenMP, [check this line](https://github.com/edubart/nelua-lang/blob/master/examples/condots.nelua#L243), it's an example on that. I also do the same [in this ray tracer](https://github.com/edubart/lunray/blob/master/lunray.nelua#L93).
>

How come I have missed that line? O.o
 
> 
> TCC doesn't do much optimizations, thus its binary output is usually very inefficient, I recommend using TCC just for quickly prototyping in a live development setup if you think GCC/Clang is being too slow to compile a big project and impacting your productivity.

So far I haven't had any problem with GCC compiling the code; actually it's on par with TCC during compilation process for some reason.

You are doing an excellent job, I have to praise you on that.
